src/DotNetty.Transport/Channels/Sockets/SocketDatagramChannel.cs

```diff
@@ -0,0 +1,481 @@
+            : this(new Socket(SocketType.Dgram, ProtocolType.Udp))
+        {
+        }
+
+        public SocketDatagramChannel(AddressFamily addressFamily)
+            : this(new Socket(addressFamily, SocketType.Dgram, ProtocolType.Udp))
+        {
+        }
+
+        public SocketDatagramChannel(Socket socket)
+            : base(null, socket)
+        {
+            this.config = new DefaultDatagramChannelConfig(this, socket);
+            this.anyRemoteEndPoint = new IPEndPoint(
+                socket.AddressFamily == AddressFamily.InterNetwork ? IPAddress.Any : IPAddress.IPv6Any,
+                IPEndPoint.MinPort);
+        }
+
+        public override IChannelConfiguration Configuration => this.config;
+
+        public override ChannelMetadata Metadata => ChannelMetadata;
+
+        protected override EndPoint LocalAddressInternal => this.Socket.LocalEndPoint;
+
+        protected override EndPoint RemoteAddressInternal => this.Socket.RemoteEndPoint;
+
+        protected override void DoBind(EndPoint localAddress)
+        {
+            this.Socket.Bind(localAddress);
+            this.CacheLocalAddress();
+
+            this.SetState(StateFlags.Active);
+        }
+
+        public override bool Active => this.Open && this.Socket.IsBound;
+
+        protected override bool DoConnect(EndPoint remoteAddress, EndPoint localAddress)
+        {
+            if (localAddress != null)
+            {
+                this.DoBind(localAddress);
+            }
+
+            bool success = false;
+            try
+            {
+                this.Socket.Connect(remoteAddress);
+                success = true;
+                return true;
+            }
+            finally
+            {
+                if (!success)
+                {
+                    this.DoClose();
+                }
+            }
+        }
+
+        protected override void DoFinishConnect(SocketChannelAsyncOperation operation)
+        {
+            throw new NotSupportedException();
+        }
+
+        protected override void DoDisconnect() => this.DoClose();
+
+        protected override void DoClose()
+        {
+            if (this.TryResetState(StateFlags.Open | StateFlags.Active))
+            {
+                this.Socket.Shutdown(SocketShutdown.Both);
+                this.Socket.Dispose();
+            }
+        }
+
+        protected override void ScheduleSocketRead()
+        {
+            SocketChannelAsyncOperation operation = this.ReadOperation;
+            operation.RemoteEndPoint = this.anyRemoteEndPoint;
+
+            IRecvByteBufAllocatorHandle handle = this.Unsafe.RecvBufAllocHandle;
+            IByteBuffer buffer = handle.Allocate(this.config.Allocator);
+            handle.AttemptedBytesRead = buffer.WritableBytes;
+            operation.UserToken = buffer;
+
+            ArraySegment<byte> bytes = buffer.GetIoBuffer();
+            operation.SetBuffer(bytes.Array, bytes.Offset, buffer.WritableBytes);
```

Comment:
technically this is not allowed: GetIoBuffer without parameters requests for ArraySegment that is `ReadableBytes` long. pls change to `GetIoBuffer(0, buffer.WritableBytes)` and use `bytes.Count` for length.